File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
    void generator(int o, int c, string seed, vector<string> &res) {
        if (o == 0 && c == 0) {
            res.push_back(seed);
            return;
        }
        
        if (o < c)
            generator(o, c - 1, seed + ')', res);
        if (o > 0)
            generator(o - 1, c, seed + '(', res);
        
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        generator(n,n,"",res);
        return res;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.rs
impl Solution {
    fn helper(o: i32, c: i32, str: String, res: &mut Vec<String>) {
        if o == 0 && c == 0 {
            res.push(str);
            return;
        }
        
        if c > o {
            Solution::helper(o, c - 1, str.clone() + ")", res);
        }
        
        if o > 0 {
            Solution::helper(o - 1, c, str + "(", res);
        }
    }
    
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut res:Vec<String> = Vec::new();
        Solution::helper(n, n, String::new(), &mut res);
        res
    }
}